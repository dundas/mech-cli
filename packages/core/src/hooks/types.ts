/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * Hook types matching Claude Code hook system
 * @see https://docs.claude.com/en/docs/claude-code/hooks
 */
export enum HookType {
  PreToolUse = 'PreToolUse',
  PostToolUse = 'PostToolUse',
  Notification = 'Notification',
  UserPromptSubmit = 'UserPromptSubmit',
  Stop = 'Stop',
  SubagentStop = 'SubagentStop',
  PreCompact = 'PreCompact',
  SessionStart = 'SessionStart',
  SessionEnd = 'SessionEnd',
}

/**
 * Hook execution result
 */
export interface HookResult {
  /** Exit code from hook (0 = success, 2 = blocking error) */
  exitCode: number;
  /** Standard output from hook */
  stdout: string;
  /** Standard error from hook */
  stderr: string;
  /** Whether the hook requests to block the operation */
  shouldBlock: boolean;
  /** Optional JSON response from hook */
  response?: HookResponse;
}

/**
 * Advanced hook response format (optional JSON output)
 */
export interface HookResponse {
  /** Whether to block the operation */
  block?: boolean;
  /** Message to display to user */
  message?: string;
  /** Additional context data */
  context?: Record<string, unknown>;
}

/**
 * Hook input data (sent via stdin as JSON)
 */
export interface HookInput {
  /** Session identifier */
  sessionId: string;
  /** Hook type being executed */
  hookType: HookType;
  /** Tool name (for tool-related hooks) */
  toolName?: string;
  /** Tool parameters (for tool-related hooks) */
  toolParams?: Record<string, unknown>;
  /** Tool result (for PostToolUse hooks) */
  toolResult?: Record<string, unknown>;
  /** User prompt (for UserPromptSubmit hooks) */
  userPrompt?: string;
  /** Additional event-specific data */
  eventData?: Record<string, unknown>;
  /** Current working directory */
  cwd: string;
  /** Project root directory */
  projectRoot: string;
  /** Timestamp */
  timestamp: number;
}

/**
 * Hook command configuration
 */
export interface HookCommand {
  /** Command type (only 'command' supported for now) */
  type: 'command';
  /** Shell command to execute */
  command: string;
  /** Optional timeout in milliseconds (default: 60000) */
  timeout?: number;
}

/**
 * Hook matcher configuration
 */
export interface HookMatcher {
  /** Regex pattern to match tool names (or '*' for all tools) */
  matcher: string;
  /** List of hook commands to execute */
  hooks: HookCommand[];
}

/**
 * Hooks configuration (mirroring Claude Code format)
 */
export interface HooksConfig {
  PreToolUse?: HookMatcher[];
  PostToolUse?: HookMatcher[];
  Notification?: HookMatcher[];
  UserPromptSubmit?: HookMatcher[];
  Stop?: HookMatcher[];
  SubagentStop?: HookMatcher[];
  PreCompact?: HookMatcher[];
  SessionStart?: HookMatcher[];
  SessionEnd?: HookMatcher[];
}

/**
 * Default hook timeout (60 seconds, matching Claude Code)
 */
export const DEFAULT_HOOK_TIMEOUT = 60000;

/**
 * Blocking exit code (exit code 2 means block)
 */
export const BLOCKING_EXIT_CODE = 2;

/**
 * Success exit code
 */
export const SUCCESS_EXIT_CODE = 0;
