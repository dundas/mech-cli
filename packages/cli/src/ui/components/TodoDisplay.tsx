/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { useUIState } from '../contexts/UIStateContext.js';
import { Box, Text } from 'ink';
import { theme } from '../semantic-colors.js';
import type { HistoryItemToolGroup } from '../types.js';
import { useMemo } from 'react';

export const TodoDisplay = () => {
  const { history } = useUIState();

  const inProgressTodoDescription = useMemo(() => {
    if (!history) {
      return null;
    }
    for (let i = history.length - 1; i >= 0; i--) {
      const item = history[i];
      if (item.type !== 'tool_group') {
        continue;
      }
      for (const tool of (item as HistoryItemToolGroup).tools) {
        if (
          tool.name !== 'Write Todos' ||
          typeof tool.resultDisplay !== 'string'
        ) {
          continue;
        }
        for (const line of tool.resultDisplay.split('\n')) {
          const match = line.match(/\[in_progress\]\\s*(.*)/);
          if (match) {
            return match[1];
          }
        }
        console.debug('in progress todo not found!');
        // If we found a Write Todos tool but no in_progress line, don't look further back.
        return null;
      }
    }
    return null;
  }, [history]);

  if (!inProgressTodoDescription) {
    return null;
  }

  return (
    <Box
      borderStyle="round"
      borderColor={theme.border.default}
      flexDirection="column"
      paddingX={1}
      borderBottom={false}
    >
      <Text>In Progress ToDo: {inProgressTodoDescription}</Text>
    </Box>
  );
};
