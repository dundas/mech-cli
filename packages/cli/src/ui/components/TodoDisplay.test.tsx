/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render } from 'ink-testing-library';
import { TodoDisplay } from './TodoDisplay.js';
import { useUIState, type UIState } from '../contexts/UIStateContext.js';
import { ToolCallStatus, type HistoryItem } from '../types.js';

// Mock the useUIState hook
vi.mock('../contexts/UIStateContext.js', () => ({
  useUIState: vi.fn(),
}));

const mockUseUIState = vi.mocked(useUIState);

const createMockUIState = (history: UIState['history'] = []): UIState =>
  ({
    history,
  }) as UIState;

describe('<TodoDisplay />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders nothing when history is empty', () => {
    mockUseUIState.mockReturnValue(createMockUIState());
    const { lastFrame } = render(<TodoDisplay />);
    expect(lastFrame()).toBe('');
  });

  it('renders nothing when no "Write Todos" tool is in history', () => {
    const history = [
      {
        id: 1,
        type: 'tool_group',
        tools: [
          {
            callId: '1',
            name: 'read_file',
            description: 'Read a file',
            status: ToolCallStatus.Success,
            resultDisplay: 'file content',
          },
        ],
      },
    ] as unknown as HistoryItem[];
    mockUseUIState.mockReturnValue(createMockUIState(history));
    const { lastFrame } = render(<TodoDisplay />);
    expect(lastFrame()).toBe('');
  });

  it('renders nothing when "Write Todos" tool has no in_progress item', () => {
    const history = [
      {
        id: 1,
        type: 'tool_group',
        tools: [
          {
            callId: '1',
            name: 'Write Todos',
            description: 'Write a todo list',
            status: ToolCallStatus.Success,
            resultDisplay: `
- [pending] Task 1
- [completed] Task 2
`,
          },
        ],
      },
    ] as unknown as HistoryItem[];
    mockUseUIState.mockReturnValue(createMockUIState(history));
    const { lastFrame } = render(<TodoDisplay />);
    expect(lastFrame()).toBe('');
  });

  it('renders the in_progress todo item and matches snapshot', () => {
    const history = [
      {
        id: 1,
        type: 'tool_group',
        tools: [
          {
            callId: '1',
            name: 'Write Todos',
            description: 'Write a todo list',
            status: ToolCallStatus.Success,
            resultDisplay: `
- [pending] Task A
- [in_progress] Task B
- [completed] Task C
`,
          },
        ],
      },
    ] as unknown as HistoryItem[];
    mockUseUIState.mockReturnValue(createMockUIState(history));
    const { lastFrame } = render(<TodoDisplay />);
    expect(lastFrame()).toMatchSnapshot();
  });

  it('renders the latest in_progress todo when multiple "Write Todos" tools exist', () => {
    const history = [
      {
        id: 1,
        type: 'tool_group',
        tools: [
          {
            callId: '1',
            name: 'Write Todos',
            description: 'Old todo list',
            status: ToolCallStatus.Success,
            resultDisplay: `
- [in_progress] Old Task
`,
          },
        ],
      },
      {
        id: 2,
        type: 'tool_group',
        tools: [
          {
            callId: '2',
            name: 'Write Todos',
            description: 'New todo list',
            status: ToolCallStatus.Success,
            resultDisplay: `
- [pending] New Task 1
- [in_progress] New Task 2
`,
          },
        ],
      },
    ] as unknown as HistoryItem[];
    mockUseUIState.mockReturnValue(createMockUIState(history));
    const { lastFrame } = render(<TodoDisplay />);
    expect(lastFrame()).toMatchSnapshot();
  });
});
