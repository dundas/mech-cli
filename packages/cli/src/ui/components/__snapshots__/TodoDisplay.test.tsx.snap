// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TodoDisplay /> > renders the in_progress todo item and matches snapshot 1`] = `""`;

exports[`<TodoDisplay /> > renders the latest in_progress todo when multiple "Write Todos" tools exist 1`] = `""`;
